### PR TITLE
error handling in minimal dispatch

### PR DESCRIPTION
**Pull request by bartlomieju**
Towards #3117 

A little prototype of error handling in minimal dispatch. I expect it to be slow but I want to start a discussion.

CC @ry

### CODE REVIEW

**Comment by ry** [comment]
@bartlomieju Does this have an effect on the deno_tcp http req/sec benchmark?

**Comment by bartlomieju** [comment]
@ry I get around 88k req/sec instead of 95k. I need to work on it a bit more, started getting some panics locally :/

**Comment by bartlomieju** [comment]
this PR 
```
  "req_per_sec": {
    "deno_tcp": 50138, 
    "deno_core_single": 62672, 
    "node_proxy_tcp": 14730, 
    "deno_proxy_tcp": 14773, 
    "deno_core_multi": 61024, 
    "deno_tcp_current_thread": 47704, 
    "hyper": 54209, 
    "node_tcp": 33282, 
    "deno_http": 13311, 
    "node_http": 17620, 
    "deno_proxy": 1014, 
    "node_proxy": 2683
  }, 
  "max_latency": {
    "deno_tcp": 8.26, 
    "deno_core_single": 1.31, 
    "node_proxy_tcp": 3.06, 
    "deno_proxy_tcp": 3.82, 
    "deno_core_multi": 5.49, 
    "deno_tcp_current_thread": 1.96, 
    "hyper": 3.36, 
    "node_tcp": 2.72, 
    "deno_http": 31.25, 
    "node_http": 4.82, 
    "deno_proxy": 18.39, 
    "node_proxy": 9.86
  },
```

master 
```
  "req_per_sec": {
    "deno_tcp": 55549, 
    "deno_core_single": 59915, 
    "node_proxy_tcp": 15679, 
    "deno_proxy_tcp": 18249, 
    "deno_core_multi": 62742, 
    "deno_tcp_current_thread": 55106, 
    "hyper": 54423, 
    "node_tcp": 35625, 
    "deno_http": 13972, 
    "node_http": 17561, 
    "deno_proxy": 1096, 
    "node_proxy": 2846
  }, 
  "max_latency": {
    "deno_tcp": 5.32, 
    "deno_core_single": 1.16, 
    "node_proxy_tcp": 2.06, 
    "deno_proxy_tcp": 3.29, 
    "deno_core_multi": 6.15, 
    "deno_tcp_current_thread": 2.04, 
    "hyper": 2.66, 
    "node_tcp": 3.26, 
    "deno_http": 38.37, 
    "node_http": 3.76, 
    "deno_proxy": 16.86, 
    "node_proxy": 9.06
  },
```

**Comment by bartlomieju** [comment]
@ry, how about we land this PR so we can remove panics for malformed control buffers and work on any potential perf problems down the road?

File: cli/js/dispatch_minimal.ts [hunk 1]
```diff
@@ -17,23 +21,57 @@ export interface RecordMinimal {
...
-    promiseId: buf32[0],
+    promiseId,
     opId,
-    arg: buf32[1],
-    result: buf32[2]
+    arg,
+    result,
+    err
   };
 }
```
Comment (ry) [inline cli/js/dispatch_minimal.ts]: This code is very much in the hot path for our most important benchmark. I'm all for landing this to see what happens, but I will revert if it causes a noticeable slow down.

Comment (bartlomieju) [reply cli/js/dispatch_minimal.ts]: Yes I agree completely. I will try to optimize it further, but we can't leave out proper error handling (which is long overdue anyway)

**Comment by ry** [comment]
@bartlomieju Sounds good - please add a unit test for the method I pointed out and I'll land it.

File: cli/js/dispatch_minimal.ts [hunk 2]
```diff
@@ -17,23 +21,57 @@ export interface RecordMinimal {
...
 export function recordFromBufMinimal(
   opId: number,
-  buf32: Int32Array
+  ui8: Uint8Array
 ): RecordMinimal {
-  if (buf32.length != 3) {
-    throw Error("Bad message");
+  const promiseId = readInt32LittleEndian(ui8, 0);
+  const arg = readInt32LittleEndian(ui8, 4);
+  const result = readInt32LittleEndian(ui8, 8);
```
Comment (ry) [inline cli/js/dispatch_minimal.ts]: This is likely too much work to decode. Converting to Int32Array and reading elements is faster.

**Comment by bartlomieju** [comment]
@ry it looks like sccache is acting up on Windows. Did you have any problems like that recently?

**Comment by bartlomieju** [comment]
@ry ready

File: cli/ops/dispatch_minimal.rs [hunk 3]
```diff
@@ -31,6 +33,44 @@ impl Into<Buf> for Record {
...
+    32, 32,
+  ];
+  let err_record = ErrorRecord {
+    promise_id: 1,
+    arg: -1,
+    error_code: 10,
+    error_message: "Error".to_string().as_bytes().to_owned(),
+  };
+  let buf: Buf = err_record.into();
+  assert_eq!(buf, expected.into_boxed_slice());
```
Comment (ry) [inline cli/ops/dispatch_minimal.rs]: 👍

**Comment by kevinkassimo** [comment]
The slowdown from this PR seems quite apparent from benchmark:

Screenshot of Deno

**Pull request merged**